File: src/main/java/epicsquid/superiorshields/SuperiorShields.java
package epicsquid.superiorshields;

import com.tterrag.registrate.Registrate;
import epicsquid.superiorshields.capability.shield.IShieldCapability;
import epicsquid.superiorshields.config.Config;
import epicsquid.superiorshields.enchantment.ModEnchantments;
import epicsquid.superiorshields.gui.GuiShieldOverlay;
import epicsquid.superiorshields.item.ModItems;
import epicsquid.superiorshields.lang.ModLang;
import epicsquid.superiorshields.network.NetworkHandler;
import epicsquid.superiorshields.setup.ModSetup;
import epicsquid.superiorshields.setup.compat.ArsCompat;
import epicsquid.superiorshields.setup.compat.BotaniaCompat;
import epicsquid.superiorshields.setup.compat.MalumCompat;
import epicsquid.superiorshields.setup.compat.ThermalCompat;
import epicsquid.superiorshields.tags.ModTags;
import net.minecraft.data.DataGenerator;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.inventory.InventoryMenu;
import net.minecraft.world.item.CreativeModeTab;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.client.event.TextureStitchEvent;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.common.capabilities.RegisterCapabilitiesEvent;
import net.minecraftforge.common.data.ForgeBlockTagsProvider;
import net.minecraftforge.common.util.Lazy;
import net.minecraftforge.eventbus.api.EventPriority;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.ModLoadingContext;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.config.ModConfig;
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
import net.minecraftforge.fml.event.lifecycle.InterModEnqueueEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import net.minecraftforge.forge.event.lifecycle.GatherDataEvent;

import javax.annotation.Nonnull;

@Mod(SuperiorShields.MODID)
public class SuperiorShields {

	public static final String MODID = "superiorshields";

	private static final Lazy<Registrate> REGISTRATE = Lazy.of(() -> Registrate.create(MODID));
	public static final String SHIELD_CURIO = "superior_shield";

	public static final CreativeModeTab ITEM_GROUP = new CreativeModeTab(SuperiorShields.MODID) {
		@Override
		@Nonnull
		public ItemStack makeIcon() {
			return new ItemStack(ModItems.IRON_SHIELD.get());
		}
	};

	public SuperiorShields() {
		ModLoadingContext.get().registerConfig(ModConfig.Type.SERVER, Config.SHIELD_SPEC, "superior-shields-common.toml");
		FMLJavaModLoadingContext.get().getModEventBus().addListener(this::setup);
		FMLJavaModLoadingContext.get().getModEventBus().addListener(this::sendImc);
		FMLJavaModLoadingContext.get().getModEventBus().addListener(this::registerCaps);
		MinecraftForge.EVENT_BUS.register(this);

		BotaniaCompat.init();
		MalumCompat.init();
		ThermalCompat.init();
		ArsCompat.init();

		ModItems.classload();
		ModEnchantments.classload();
		ModLang.classload();

		FMLJavaModLoadingContext.get().getModEventBus().addListener(EventPriority.LOWEST, this::gatherData);
	}

	public void setup(FMLCommonSetupEvent event) {
		MinecraftForge.EVENT_BUS.register(new EventManager());
		NetworkHandler.register();
	}

	@SubscribeEvent
	public void registerCaps(final RegisterCapabilitiesEvent event) {
		event.register(IShieldCapability.class);
	}

	public void gatherData(GatherDataEvent event) {
		DataGenerator generator = event.getGenerator();
		if (event.includeServer()) {
			ForgeBlockTagsProvider b = new ForgeBlockTagsProvider(generator, event.getExistingFileHelper());
			generator.addProvider(new ModTags(generator, b, event.getExistingFileHelper()));
		}
	}

	public void sendImc(InterModEnqueueEvent event) {
		ModSetup.sendImc();
	}

	public static Registrate registrate() {
		return REGISTRATE.get();
	}

	@Mod.EventBusSubscriber(modid = MODID, value = Dist.CLIENT, bus = Mod.EventBusSubscriber.Bus.MOD)
	public static class ClientProxy {
		@SubscribeEvent
		public static void stitchTextures(TextureStitchEvent.Pre event) {
			if (event.getAtlas().location() == InventoryMenu.BLOCK_ATLAS) {
				event.addSprite(new ResourceLocation(SuperiorShields.MODID, "item/empty_shield_slot"));
			}
		}

		@SubscribeEvent
		public static void setupClient(FMLClientSetupEvent event) {
//			MinecraftForge.EVENT_BUS.register(new ClientEventManager());
			GuiShieldOverlay.init();
		}
	}
}

File: src/main/kotlin/dev/epicsquid/superiorshields/SuperiorShields.kt
package dev.epicsquid.superiorshields

import com.tterrag.registrate.Registrate
import dev.epicsquid.superiorshields.config.Config
import dev.epicsquid.superiorshields.data.SuperiorShieldsTags
import dev.epicsquid.superiorshields.event.ClientEventManager
import dev.epicsquid.superiorshields.event.EventManager
import dev.epicsquid.superiorshields.network.NetworkHandler
import dev.epicsquid.superiorshields.registry.*
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.item.CreativeModeTab
import net.minecraft.world.item.ItemStack
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.common.data.ForgeBlockTagsProvider
import net.minecraftforge.data.event.GatherDataEvent
import net.minecraftforge.eventbus.api.EventPriority.LOWEST
import net.minecraftforge.fml.InterModComms
import net.minecraftforge.fml.ModLoadingContext
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.config.ModConfig
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent
import net.minecraftforge.fml.event.lifecycle.InterModEnqueueEvent
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext
import thedarkcolour.kotlinforforge.forge.runForDist
import top.theillusivec4.curios.api.CuriosApi
import top.theillusivec4.curios.api.SlotTypeMessage
import top.theillusivec4.curios.api.SlotTypeMessage.Builder

@Mod(SuperiorShields.MODID)
class SuperiorShields {
	companion object {
		const val MODID = "superiorshields"
		const val SUPERIOR_SHIELD_CURIO = "superior_shield"

		val tab = object : CreativeModeTab(MODID) {
			override fun makeIcon() = ItemStack(ItemRegistry.ironShield)
		}

		val registrate: Registrate by lazy {
			Registrate.create(MODID).creativeModeTab { tab }
		}
	}

	init {
		ModLoadingContext.get()
			.registerConfig(ModConfig.Type.SERVER, Config.SHIELDS_CONFIG_SPEC, "superior-shields-server.toml")
		val modEventBus = FMLJavaModLoadingContext.get().modEventBus

		MinecraftForge.EVENT_BUS.register(EventManager)
		modEventBus.addListener { _: FMLCommonSetupEvent -> NetworkHandler.register() }
		modEventBus.addListener { _: InterModEnqueueEvent -> onInterModEnqueue() }
		runForDist(
			clientTarget = { modEventBus.register(ClientEventManager) },
			serverTarget = {}
		)

		EnchantmentRegistry.classload()
		ItemRegistry.classload()
		LangRegistry.classload()

		FMLJavaModLoadingContext.get().modEventBus.addListener(LOWEST, this::onGatherData)
	}

	private fun onInterModEnqueue() {
		InterModComms.sendTo(
			CuriosApi.MODID, SlotTypeMessage.REGISTER_TYPE
		) {
			Builder(SUPERIOR_SHIELD_CURIO).apply {
				priority(220)
				icon(ResourceLocation(MODID, "item/empty_shield_slot"))
			}.build()
		}
	}

	private fun onGatherData(event: GatherDataEvent) {
		val generator = event.generator
		val blockTagsProvider = ForgeBlockTagsProvider(generator, event.existingFileHelper)
		generator.addProvider(
			event.includeServer(),
			SuperiorShieldsTags(generator, blockTagsProvider, event.existingFileHelper)
		)
	}
}